response.go

```diff
@@ -103,185 +176,444 @@ func (info *SQLInfo) DecodeMsgpack(d *msgpack.Decoder) error {
 	return nil
 }
 
-func (resp *Response) smallInt(d *msgpack.Decoder) (i int, err error) {
-	b, err := resp.buf.ReadByte()
+func smallInt(d *msgpack.Decoder, buf *smallBuf) (i int, err error) {
+	b, err := buf.ReadByte()
 	if err != nil {
 		return
 	}
 	if b <= 127 {
 		return int(b), nil
 	}
-	resp.buf.UnreadByte()
+	buf.UnreadByte()
 	return d.DecodeInt()
 }
 
-func (resp *Response) decodeHeader(d *msgpack.Decoder) (err error) {
+func decodeHeader(d *msgpack.Decoder, buf *smallBuf) (Header, error) {
 	var l int
-	d.Reset(&resp.buf)
+	var err error
+	d.Reset(buf)
 	if l, err = d.DecodeMapLen(); err != nil {
-		return
+		return Header{}, err
 	}
+	decodedHeader := Header{}
 	for ; l > 0; l-- {
 		var cd int
-		if cd, err = resp.smallInt(d); err != nil {
-			return
+		if cd, err = smallInt(d, buf); err != nil {
+			return Header{}, err
 		}
 		switch iproto.Key(cd) {
 		case iproto.IPROTO_SYNC:
 			var rid uint64
 			if rid, err = d.DecodeUint64(); err != nil {
-				return
+				return Header{}, err
 			}
-			resp.RequestId = uint32(rid)
+			decodedHeader.RequestId = uint32(rid)
 		case iproto.IPROTO_REQUEST_TYPE:
-			var rcode uint64
-			if rcode, err = d.DecodeUint64(); err != nil {
-				return
+			var rcode int
+			if rcode, err = d.DecodeInt(); err != nil {
+				return Header{}, err
 			}
-			resp.Code = uint32(rcode)
+			decodedHeader.Code = rcode
 		default:
 			if err = d.Skip(); err != nil {
-				return
+				return Header{}, err
 			}
 		}
 	}
+	return decodedHeader, nil
+}
+
+type decodeInfo struct {
+	stmtID             uint64
+	bindCount          uint64
+	serverProtocolInfo ProtocolInfo
+	errorExtendedInfo  *BoxError
+
+	decodedError string
+}
+
+func (info *decodeInfo) parseData(resp *baseResponse) error {
+	if info.stmtID != 0 {
+		stmt := &Prepared{
+			StatementID: PreparedID(info.stmtID),
+			ParamCount:  info.bindCount,
+		}
+		resp.data = []interface{}{stmt}
+	}
+
+	// Tarantool may send only version >= 1
+	if info.serverProtocolInfo.Version != ProtocolVersion(0) ||
+		info.serverProtocolInfo.Features != nil {
+		if info.serverProtocolInfo.Version == ProtocolVersion(0) {
+			return fmt.Errorf("no protocol version provided in Id response")
+		}
+		if info.serverProtocolInfo.Features == nil {
+			return fmt.Errorf("no features provided in Id response")
+		}
+		resp.data = []interface{}{info.serverProtocolInfo}
+	}
 	return nil
```

Comment:
It is unexpected that it could be in the same time. So it's better to make it clear (or using `switch`).
```suggestion
	if info.stmtID != 0 {
		stmt := &Prepared{
			StatementID: PreparedID(info.stmtID),
			ParamCount:  info.bindCount,
		}
		resp.data = []interface{}{stmt}
		return nil
	}

	// Tarantool may send only version >= 1
	if info.serverProtocolInfo.Version != ProtocolVersion(0) ||
		info.serverProtocolInfo.Features != nil {
		if info.serverProtocolInfo.Version == ProtocolVersion(0) {
			return fmt.Errorf("no protocol version provided in Id response")
		}
		if info.serverProtocolInfo.Features == nil {
			return fmt.Errorf("no features provided in Id response")
		}
		resp.data = []interface{}{info.serverProtocolInfo}
		return nil
	}
	return nil
```